File: Components/Strata.Foundation.Core/main/typescript/Utility/ExpendableContext.ts
export
class ExpendableContext<T>
{
    private readonly value: T;
    private readonly allowed: number;
    private          remaining: number;

    public constructor(value:T, limit:number)
    {
        this.value = value;
        this.allowed = limit;
        this.remaining = limit;
    }

    public decrementRemaining(): ExpendableContext<T>
    {
        --this.remaining;
        return this;
    }

    public getValue(): T
    {
        return this.value;
    }

    public getAllowed(): number
    {
        return this.allowed;
    }

    public getRemaining(): number
    {
        return this.remaining;
    }

    public isExpended(): boolean
    {
        return this.remaining <= 0;
    }

}

File: Components/Strata.Foundation.Core/main/java/strata/foundation/core/utility/ExpendableContext.java
/// ///////////////////////////////////////////////////////////////////////////
// ExpendableContext.java
//////////////////////////////////////////////////////////////////////////////

package strata.foundation.core.utility;

import java.io.Serializable;

class ExpendableContext<T>
    implements Serializable
{
    private final T   value;
    private final int allowed;
    private int       remaining;

    public
    ExpendableContext(T value,int allowed)
    {
        this.value = value;
        this.allowed = allowed;
        this.remaining = this.allowed;
    }

    public ExpendableContext<T>
    decrementRemaining()
    {
        --remaining;
        return this;
    }

    public T
    getValue() { return value; }

    public int
    getAllowed() { return allowed; }

    public int
    getRemaining() { return remaining; }

    public boolean
    isExpended()
    {
        return remaining <= 0;
    }
}

//////////////////////////////////////////////////////////////////////////////
